### ISSUE TITLE
20 * 2000 TMXTiledMap last few tiles can't display

### ISSUE THREAD

**Issue by folgerhu** [cocos2d/cocos2d-x#12994]
cocos2d-x v3.7

Map size: 20 x 2000
Last few tiles can't display in cocos2d-x:

``` cpp
    auto map = TMXTiledMap::create("map.tmx");
    addChild(map, 0, kTagTileMap);
```

<img width="1279" alt="screen shot 2015-07-24 at 11 52 05 am" src="https://cloud.githubusercontent.com/assets/3455415/8867323/771d0bf2-31fa-11e5-87f4-7da16bc74791.png">

Here is my map resource: http://vdisk.weibo.com/s/u2pFxSLc-nv-

**Comment by stevetranby** [comment]
I believe this issue is related overflow on the OpenGL vertex buffer or index buffer, but I haven't nailed it down quite yet. I do know that the maximum # of tiles with the current TMX system is 128*128=16384 and the # of visible tiles in your .tmx map is something around 16390 or so and is thus cutting off the last 6 visible tiles (approx).

I'll respond with more as I look at it. I'm looking to improve TMX tiled map.

Edit: the number is based on storing information in an unsigned short somewhere in the renderer or OpenGL. 2^16 / 4 = 16384 (4 bytes per tile => 16384 max visible tiles). It's related to the fact that TMX uses a SpriteBatch as renderer and the atlas index overflows at this number related to the amount of visible tiles.

**Comment by stevetranby** [comment]
The real solution is likely to enable support for splitting up a map into layer chunks. Could create multiple vertex/index buffers as well, but might as well support a better abstraction with splitting up a map into max sized regions by chunking as needed. This lends itself as well toward streaming in map and loading only adjacent chunks as player moves around map. If you need to see the full map of such a large size one could still load and render all chunks with the limits of CPU/GPU memory size (and performance for FPS) instead, thus the limits would be on the rendering side and not the storage (map representation).

**Comment by dabingnn** [comment]
Yes, Map streaming sounds to be a good solution for the limitation of vertex and index buffer。
